src/hotspot/share/opto/rangeinference.hpp

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_OPTO_RANGEINFERENCE_HPP
+#define SHARE_OPTO_RANGEINFERENCE_HPP
+
+#include "utilities/globalDefinitions.hpp"
+#include <type_traits>
+
+class outputStream;
+class Type;
+class TypeInt;
+class TypeLong;
+
+template <class T>
+class RangeInt {
+public:
+  T _lo;
+  T _hi;
+};
+
+/**
+ * Bits that are known to be 0 or 1. A value v satisfies this constraint iff
+ * (v & zeros) == 0 && (~v & ones) == 0. I.e, all bits that is set in zeros
+ * must be unset in v, and all bits that is set in ones must be set in v.
```

Comment:
That is quite counter-intuitive. Is there a good reason for this?
I would have expected that `zero[i] = 1` would mean that a zero is allowed, and `ones[i] = 1` that a one is allowed.

Basically, when I see `zero[i]` I expect it to be a boolean that answers me this question: "can it be a zero?". But you are telling me I'm supposed to ask "Must it not be a zero"?

You are telling me that `zero[i] = 1` and `ones[i] = 0` means that it must be a `1`.

I know that changing it now would be a lot of effort. But the risk of being unintuitive is that even less people can quickly fix bugs in this code.

@vnkozlov what do you think about this?

Reply:
You are a bit confused, the `zeros` and `ones` answer the question: Must this bit is 0 (or 1). Which means that `zero[i] = 1` means that the bit must be a `0`.

Reply:
The condition could be converted to `(v & ones) == ones && (~v & zeros) == 0`. 
It is matching your statement and easy to understand. I would let C++ compiler to optimize it instead of encoding it in confusing way.

Reply:
That's a good idea, I have made that change.